### PR TITLE
Cleanup Special Matrices code

### PR DESCRIPTION
**Pull request by alyst**
Simplifies a bit the code to generate *elimination*, *duplication* and *commutation* matrices.
The latter is implemented as a special `CommutationMatrix <: AbstractMatrix{Int}` subtype, which provides elegant integration into LinearAlgebra API.

I was actually trying to check, whether there are Julia packages implementing these special matrices.
So far I have found [*VectorTransformations.jl*](https://github.com/tbeason/VectorizationTransformations.jl) package, but it works by constructing sparse matrices with the predefined structure, not by implementing custom code that does the expected transformation directly (which might be faster).

Cherry-picked from #193

### CODE REVIEW

File: src/loss/ML/FIML.jl [hunk 1]
```diff
@@ -37,7 +37,7 @@ mutable struct SemFIML{INV, C, L, O, M, IM, I, T, U, W} <: SemLossFunction
 
     mult::T
 
-    commutation_indices::U
+    commutator::CommutationMatrix
```
Comment (Maximilian-Stefan-Ernst) [inline src/loss/ML/FIML.jl:40]: I think you have to delete the type parameter U from the signature of the FIML type.

File: src/additional_functions/commutation_matrix.jl [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# transpose linear indices of the square n×n matrix
+# i.e.
+# 1 4
+# 2 5 =>  1 2 3
+# 3 6     4 5 6
```
Comment (Maximilian-Stefan-Ernst) [inline src/additional_functions/commutation_matrix.jl:5]:
```suggestion
"""
transpose linear indices of the n×m matrix
i.e.
1 4
2 5 =>  1 2 3
3 6     4 5 6
"""
```

**Review by Maximilian-Stefan-Ernst — COMMENTED** [review]
Nice, this is ready to merge once the tests pass. I have added some unit tests for the special matrices.

**Comment by Maximilian-Stefan-Ernst** [comment]
Thanks for the additional tests. Should we merge now? It seems like formatter still complains...

**Comment by alyst** [comment]
It looks ready to me.
I have no idea why the formatter complains: the diff looks identical, I don't see any trailing spaces or anything like that. But if there would be any issues upon merging, should be easy to fix.

**Comment by alyst** [comment]
Thank you for reviewing and merging! I'm preparing a next portion of commits to cherry-pick, but it would have to wait a day or two.

**Pull request merged**